x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptor.java

```diff
@@ -137,6 +190,142 @@ public <T extends TransportResponse> void sendRequest(
                         sendWithUser(connection, action, request, options, handler, sender);
                     }
             }
+
+            private void assertNoRemoteAccessHeadersInContext() {
+                assert securityContext.getThreadContext().getHeader(REMOTE_ACCESS_CLUSTER_CREDENTIAL_HEADER_KEY) == null
+                    : "remote access headers should not be in security context";
+                assert securityContext.getThreadContext()
+                    .getHeader(RemoteAccessAuthentication.REMOTE_ACCESS_AUTHENTICATION_HEADER_KEY) == null
+                    : "remote access headers should not be in security context";
+            }
```

Comment:
I tried to scope things as tightly as possible to keep `SecurityServerTransportInterceptor` de-cluttered -- this method is only relevant in the context of `AsyncSender` and including it here (instead of on `SecurityServerTransportInterceptor`) excludes it from the list of methods listed for `SecurityServerTransportInterceptor` by the IDE (in Intellij it's CMD + F12)